File: template/capabilities/_env/contract.ts
#!/usr/bin/env tsx
/**
 * capabilities/_env/contract.ts — the capability CONTRACT (plan P0.9, GAP-4).
 *
 * "Build the protocol before the workers." Every capability — TS or (via the
 * Python mirror `contract.py`) Python — speaks this one shape so the future
 * orchestration spine (P2) can drive them uniformly:
 *
 *   - a structured RESULT ENVELOPE on stdout  ({ success, capability, outputs[], metrics{}, ... })
 *   - writes to the disposable WORK DIR        (out/work/<project>/<stage>/)
 *   - appends a PROVENANCE record              (out/work/<project>/provenance.log, append-only)
 *
 * The full manifest stage-schema lands in P2 (orchestrate/); this is the lean
 * subset every P1 engine needs now. Keep it dependency-free (runs under tsx).
 */
import { spawnSync } from 'node:child_process';
import * as crypto from 'node:crypto';
import * as fs from 'node:fs';
import * as path from 'node:path';

/** Repo root = two levels up from capabilities/_env/. */
export const REPO_ROOT = path.resolve(__dirname, '..', '..');

/**
 * The OPTIONAL Python venv's interpreter (created by `_env/setup-venv.ts`), per-OS.
 * Engines that need it must degrade gracefully when it does not exist (R-D).
 */
export const VENV_PY =
  process.platform === 'win32'
    ? path.join(REPO_ROOT, 'capabilities', '.venv', 'Scripts', 'python.exe')
    : path.join(REPO_ROOT, 'capabilities', '.venv', 'bin', 'python');

/** The structured envelope every capability prints (last line of stdout) as JSON. */
export interface CapabilityResult<M = Record<string, unknown>> {
  success: boolean;
  capability: string; // e.g. "audio/master", "color/grade", "acquire/download-asset"
  outputs: string[]; // absolute paths this run produced
  metrics: M; // capability-specific measurements (LUFS, sceneCuts, bytes, ...)
  warnings?: string[];
  error?: string;
  startedAt: string;
  finishedAt: string;
  durationMs: number;
}

/** A provenance line: what ran, with what, producing what (output hashes). */
export interface ProvenanceRecord {
  ts: string;
  capability: string;
  args?: string[];
  outputs?: { path: string; sha256: string; bytes: number }[];
  source?: string; // for acquire/: the origin URL
  note?: string;
}

/**
 * Resolve (and create) the disposable per-stage work directory.
 * GAP-9: work/<project>/<stage>/ is disposable under out/ (regenerable).
 */
export function workDir(project: string, stage: string): string {
  const dir = path.join(REPO_ROOT, 'out', 'work', sanitize(project), sanitize(stage));
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

/** The per-project provenance log (append-only). Lives beside the work tree for P1; P2 may git-track it. */
export function provenancePath(project: string): string {
  const dir = path.join(REPO_ROOT, 'out', 'work', sanitize(project));
  fs.mkdirSync(dir, { recursive: true });
  return path.join(dir, 'provenance.log');
}

export function appendProvenance(project: string, rec: ProvenanceRecord): void {
  fs.appendFileSync(provenancePath(project), JSON.stringify(rec) + '\n', 'utf8');
}

export function sha256File(p: string): string {
  return crypto.createHash('sha256').update(fs.readFileSync(p)).digest('hex');
}

export function describeOutputs(paths: string[]): { path: string; sha256: string; bytes: number }[] {
  return paths
    .filter((p) => fs.existsSync(p) && fs.statSync(p).isFile())
    .map((p) => ({ path: p, sha256: sha256File(p), bytes: fs.statSync(p).size }));
}

/** Print the envelope as a single JSON line on stdout (machine-readable) — the LAST thing a capability emits. */
export function emit<M>(result: CapabilityResult<M>): void {
  process.stdout.write(JSON.stringify(result) + '\n');
}

/**
 * Wrap a capability body: time it, catch errors, always emit a valid envelope,
 * and exit non-zero on failure. `project` (when given) routes provenance.
 */
export async function runCapability<M = Record<string, unknown>>(
  capability: string,
  body: () => Promise<{ outputs: string[]; metrics: M; warnings?: string[]; project?: string; args?: string[]; source?: string }>,
): Promise<void> {
  const startedAt = new Date();
  try {
    const r = await body();
    const finishedAt = new Date();
    const result: CapabilityResult<M> = {
      success: true,
      capability,
      outputs: r.outputs,
      metrics: r.metrics,
      warnings: r.warnings,
      startedAt: startedAt.toISOString(),
      finishedAt: finishedAt.toISOString(),
      durationMs: finishedAt.getTime() - startedAt.getTime(),
    };
    if (r.project) {
      appendProvenance(r.project, {
        ts: finishedAt.toISOString(),
        capability,
        args: r.args,
        outputs: describeOutputs(r.outputs),
        source: r.source,
      });
    }
    emit(result);
  } catch (e) {
    const finishedAt = new Date();
    emit({
      success: false,
      capability,
      outputs: [],
      metrics: {} as M,
      error: e instanceof Error ? e.message : String(e),
      startedAt: startedAt.toISOString(),
      finishedAt: finishedAt.toISOString(),
      durationMs: finishedAt.getTime() - startedAt.getTime(),
    });
    process.exit(1);
  }
}

/** Read a model id from the single source of truth (_env/models.json). Honors an env override. */
export function modelId(dotPath: string): string {
  const models = JSON.parse(fs.readFileSync(path.join(__dirname, 'models.json'), 'utf8'));
  let node: unknown = models;
  for (const key of dotPath.split('.')) {
    node = (node as Record<string, unknown>)?.[key];
  }
  const entry = node as { id?: string; envOverride?: string } | undefined;
  if (!entry?.id) throw new Error(`models.json has no model at "${dotPath}"`);
  if (entry.envOverride && process.env[entry.envOverride]) return process.env[entry.envOverride] as string;
  return entry.id;
}

/** Validate an input path exists and is a file; throw a clear error otherwise (capabilities never trust argv blindly). */
export function requireInputFile(p: string | undefined, label = 'input'): string {
  if (!p) throw new Error(`missing ${label} path`);
  if (!fs.existsSync(p) || !fs.statSync(p).isFile()) throw new Error(`${label} not found: ${p}`);
  return path.resolve(p);
}

/** Run a child process, returning {status, stdout, stderr}; never throws on non-zero. */
export function run(cmd: string, args: string[], opts: { cwd?: string } = {}): { status: number; stdout: string; stderr: string } {
  const r = spawnSync(cmd, args, { encoding: 'utf8', maxBuffer: 64 * 1024 * 1024, cwd: opts.cwd });
  return { status: r.status ?? -1, stdout: r.stdout ?? '', stderr: r.stderr ?? '' };
}

function sanitize(s: string): string {
  return s.replace(/[^a-zA-Z0-9._-]/g, '-');
}

/**
 * Load `.env` into process.env (dependency-free — tsx does not auto-load it).
 * Looks at the repo root and cwd. Never overwrites an already-set var. Never prints values.
 */
export function loadDotEnv(): void {
  const seen = new Set<string>();
  for (const file of [path.join(REPO_ROOT, '.env'), path.join(process.cwd(), '.env')]) {
    if (seen.has(file) || !fs.existsSync(file)) continue;
    seen.add(file);
    for (const raw of fs.readFileSync(file, 'utf8').split(/\r?\n/)) {
      const line = raw.trim();
      if (!line || line.startsWith('#')) continue;
      const eq = line.indexOf('=');
      if (eq === -1) continue;
      const key = line.slice(0, eq).trim();
      let val = line.slice(eq + 1).trim();
      if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
        val = val.slice(1, -1);
      }
      if (key && process.env[key] === undefined) process.env[key] = val;
    }
  }
}

/** True if the named env var (or .env entry) is present — checks PRESENCE only, never reads/prints the value. */
export function hasEnv(name: string): boolean {
  loadDotEnv();
  return !!process.env[name];
}

File: template/capabilities/_env/contract.py
"""capabilities/_env/contract.py — the Python mirror of the capability contract (plan P0.9, GAP-4).

Python capabilities (audio/*.py, color/grade.py, ingest/*.py) speak the same shape
as the TS contract (contract.ts): a JSON RESULT ENVELOPE on stdout, writes under the
disposable WORK DIR (out/work/<project>/<stage>/), and an append-only PROVENANCE log.

Also resolves the FULL ffmpeg/ffprobe build (same order as _env/ffmpeg.ts) so Python
never falls back to a stripped binary. Dependency-free (stdlib only).
"""
from __future__ import annotations

import hashlib
import json
import os
import re
import shutil
import subprocess
import sys
import time
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable

_ENV_DIR = Path(__file__).resolve().parent
REPO_ROOT = _ENV_DIR.parent.parent
_EXE = ".exe" if sys.platform == "win32" else ""

# The OPTIONAL Python venv's own interpreter (created by _env/setup-venv.ts), per-OS.
VENV_PY = str(
    REPO_ROOT / "capabilities" / ".venv"
    / ("Scripts" if sys.platform == "win32" else "bin")
    / f"python{_EXE}"
)


def _exists_file(p: str | os.PathLike | None) -> bool:
    return bool(p) and Path(p).is_file()


def _vibe_bin(exe: str) -> str | None:
    """The provisioned per-project binary dir (.vibe/bin), checked from the project
    root this file lives in AND from cwd (scripts run with cwd = project root)."""
    for root in (REPO_ROOT, Path.cwd()):
        cand = root / ".vibe" / "bin" / exe
        if cand.is_file():
            return str(cand)
    return None


def resolve_ffmpeg() -> str:
    """ffmpeg: VIBE_FFMPEG (file or dir) -> .vibe/bin -> PATH (mirrors ffmpeg.ts)."""
    ov = os.environ.get("VIBE_FFMPEG")
    if ov:
        if _exists_file(ov):
            return ov
        cand = Path(ov) / f"ffmpeg{_EXE}"
        if cand.is_file():
            return str(cand)
    provisioned = _vibe_bin(f"ffmpeg{_EXE}")
    if provisioned:
        return provisioned
    return shutil.which("ffmpeg") or "ffmpeg"


def resolve_ffprobe() -> str:
    ov = os.environ.get("VIBE_FFPROBE")
    if ov:
        if _exists_file(ov):
            return ov
        cand = Path(ov) / f"ffprobe{_EXE}"
        if cand.is_file():
            return str(cand)
    ffmpeg = resolve_ffmpeg()
    if ffmpeg != "ffmpeg":
        sib = Path(ffmpeg).parent / f"ffprobe{_EXE}"
        if sib.is_file():
            return str(sib)
    provisioned = _vibe_bin(f"ffprobe{_EXE}")
    if provisioned:
        return provisioned
    return shutil.which("ffprobe") or "ffprobe"


def _sanitize(s: str) -> str:
    return re.sub(r"[^a-zA-Z0-9._-]", "-", s)


def work_dir(project: str, stage: str) -> Path:
    d = REPO_ROOT / "out" / "work" / _sanitize(project) / _sanitize(stage)
    d.mkdir(parents=True, exist_ok=True)
    return d


def provenance_path(project: str) -> Path:
    d = REPO_ROOT / "out" / "work" / _sanitize(project)
    d.mkdir(parents=True, exist_ok=True)
    return d / "provenance.log"


def sha256_file(p: str | os.PathLike) -> str:
    h = hashlib.sha256()
    with open(p, "rb") as f:
        for chunk in iter(lambda: f.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def describe_outputs(paths: list[str]) -> list[dict[str, Any]]:
    out = []
    for p in paths:
        if Path(p).is_file():
            out.append({"path": p, "sha256": sha256_file(p), "bytes": Path(p).stat().st_size})
    return out


def append_provenance(project: str, rec: dict[str, Any]) -> None:
    with open(provenance_path(project), "a", encoding="utf-8") as f:
        f.write(json.dumps(rec) + "\n")


def model_id(dot_path: str) -> str:
    """Read a model id from the single source of truth (_env/models.json), honoring an env override."""
    models = json.loads((_ENV_DIR / "models.json").read_text(encoding="utf-8"))
    node: Any = models
    for key in dot_path.split("."):
        node = (node or {}).get(key)
    if not node or "id" not in node:
        raise ValueError(f'models.json has no model at "{dot_path}"')
    ov = node.get("envOverride")
    if ov and os.environ.get(ov):
        return os.environ[ov]
    return node["id"]


@dataclass
class CapabilityResult:
    capability: str
    success: bool = True
    outputs: list[str] = field(default_factory=list)
    metrics: dict[str, Any] = field(default_factory=dict)
    warnings: list[str] = field(default_factory=list)
    error: str | None = None
    startedAt: str = ""
    finishedAt: str = ""
    durationMs: int = 0


def emit(result: CapabilityResult) -> None:
    """Print the envelope as a single JSON line on stdout — the LAST thing a capability emits."""
    sys.stdout.write(json.dumps(asdict(result)) + "\n")
    sys.stdout.flush()


def run_capability(capability: str, body: Callable[[], dict[str, Any]]) -> None:
    """Time `body`, catch errors, always emit a valid envelope, exit non-zero on failure.

    `body` returns a dict: {outputs, metrics, warnings?, project?, args?, source?}.
    """
    t0 = time.time()
    started = datetime.now(timezone.utc).isoformat()
    try:
        r = body() or {}
        finished = datetime.now(timezone.utc).isoformat()
        res = CapabilityResult(
            capability=capability,
            success=True,
            outputs=r.get("outputs", []),
            metrics=r.get("metrics", {}),
            warnings=r.get("warnings", []),
            startedAt=started,
            finishedAt=finished,
            durationMs=int((time.time() - t0) * 1000),
        )
        if r.get("project"):
            append_provenance(r["project"], {
                "ts": finished,
                "capability": capability,
                "args": r.get("args"),
                "outputs": describe_outputs(res.outputs),
                "source": r.get("source"),
            })
        emit(res)
    except Exception as e:  # noqa: BLE001 — the boundary that guarantees a valid envelope
        emit(CapabilityResult(
            capability=capability,
            success=False,
            error=str(e),
            startedAt=started,
            finishedAt=datetime.now(timezone.utc).isoformat(),
            durationMs=int((time.time() - t0) * 1000),
        ))
        sys.exit(1)


def ffprobe_duration(path: str) -> float:
    """Duration in seconds via the full ffprobe build."""
    out = subprocess.run(
        [resolve_ffprobe(), "-v", "quiet", "-show_entries", "format=duration",
         "-of", "default=nokey=1:noprint_wrappers=1", path],
        capture_output=True, text=True, check=True,
    ).stdout.strip()
    return float(out)
